---
title: "QC_clustering_DEA"
author: "Charles"
date: '2022-07-28'
output: html_document
editor_options: 
  chunk_output_type: console
---

## Description

This is a public data set with accession number [GSE178318](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi). It consists of a Single-Cell Atlas of Liver Metastases of Colorectal Cancer to Reveal the Reprogramming of the Tumor Microenvironment in Response to Preoperative Chemotherapy.

The goals will change progressively, but for now, one of the principal goals is to check the expression of lymphotoxins in metastasised liver cancer cells from CRC.

## Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## Loading libraries

```{r libraries}
library(Seurat)
library(SeuratObject)
library(SeuratWrappers)
library(SeuratDisk)
library(ggplot2)
library(tidyverse)
library(openxlsx)
library(Matrix)
```

## **1. Preparation for quality control**

### **1. Load in the raw data, prepare it and create the seurat object**

```{r data_loading}
# check files
list.files("rawData")

barcodes <- read.table("rawData/GSE178318_barcodes.tsv", header = FALSE, sep = "\t")
head(barcodes)

genes <- read.table("rawData/GSE178318_genes.tsv", header = FALSE, sep = "\t")
head(genes)

raw_matrix <- readMM("rawData/GSE178318_matrix.mtx")
dim(raw_matrix)

# set row names and column names
rownames(raw_matrix) <- genes$V2
colnames(raw_matrix) <- barcodes$V1

# verify row names and column names 
rownames(raw_matrix)[1:10]
colnames(raw_matrix)[1:10]

# save back the matrix
saveRDS(raw_matrix,
        "objects/raw_matrix.rds", compress = FALSE)

# to create seurat object
seurat <- CreateSeuratObject(raw_matrix,
                             min.cells = 10,
                             names.delim = "_")
rm(raw_matrix)
gc()
``` 

The data set has been loaded, gene names and column names added. But looking at the column names, it appears that the last item after the last underscore is specifying if the tissue origin of the cell, i.e. if it is CRC or not. So using string separation, I will evaluate all additional information that might be hidden in the barcode string.

### **2. Add cell tissue origin to the seurat object's metadata**.

```{r cell_origin}
barcodes <- barcodes %>%
                      separate(V1, into = c("barcode", "subjectID", "tissue_origin"), sep = "_")
seurat$subjectID <- barcodes$subjectID
seurat$group <- barcodes$tissue_origin

# save the raw seurat object
saveRDS(seurat,
        "objects/raw_seurat.rds", compress = FALSE)

table(seurat$group)
```

**CRC** are the colorectal carcinoma primary tissue,
**LM** are the liver metastasis tissue
**PBMC** are the Peripheral blood mononuclear cells

Data set has been read in, prepared and seurat object sucessfully created and raw seurat object saved.

## **Perform quality control**

### **Calculate percentages of genes in each cell that map tp mitochondrial or ribosomal genes.**

```{r mito_ribo}
# Compute percent ribo and mito ratios
seurat$riboRatio <- PercentageFeatureSet(object = seurat, pattern = "^RP[SL]")
seurat$riboRatio <- seurat@meta.data$riboRatio / 100

# check success of calculation
summary(seurat$riboRatio)

seurat$mitoRatio <- PercentageFeatureSet(object = seurat, pattern = "^MT")
seurat$mitoRatio <- seurat@meta.data$mitoRatio / 100

# check success of calculation
summary(seurat$mitoRatio)
```

### **Evaluate cell count per group (CRCThere are many reasons why a career in data science is attractive. For one, it's an industry with high salaries, even when compared to other s, LM, PBMC).**

```{r cell_counts} 
# create path to the directories
figuresdir <- paste0("figures/")
figuresdir
objectsdir <- paste0("objects/")
objectsdir
resultsdir <- paste0("results/")
resultsdir 

# Visualize the number of cell counts per sample
seurat@meta.data %>% 
     	ggplot(aes(x=group, fill=group)) + 
     	geom_bar() +
     	theme_classic() +
     	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
     	      aspect.ratio = 1) +
     	theme(plot.title = element_text(hjust=0.5, face="bold")) +
     	ggtitle("NCells")
ggsave(filename=paste0(figuresdir, "/", "cell_counts_per_sample",
                       ".png"), width = 5, 
       height = 5, units = 'in', dpi = 300)
```

The results show that most of the cells in the data set are liver metastatic cells.

### **Calculate the number of UMI counts per cell in each group - with histogram.**

```{r nUMI_histo}
# Visualize the distribution of transcripts detected per cell via a histogram
seurat@meta.data %>% 
     	ggplot(aes(color=group, x=nCount_RNA, fill= group)) + 
     	geom_density(alpha = 0.2) + 
     	theme_classic() +
     	scale_x_log10() + 
     	geom_vline(xintercept = 500) +
      theme(aspect.ratio = 1)
ggsave(filename=paste0(figuresdir, "/", "genes_per_cell_final",
                       ".png"), width = 5, 
       height = 5, units = 'in', dpi = 300)
```

Most of the UMI counts are between 500 and 100000, with a peak at 3000. The UMI counts per cell should generally be above 500, that is the low end of what we expect.

### **Calculate the number of of UMI counts per cell in each group - boxplot.**

```{r nGenes_boxplot}
# Visualize the distribution of genes detected per cell via boxplot
seurat@meta.data %>% 
     	ggplot(aes(x=group, y=log10(nCount_RNA), fill=group)) + 
     	geom_boxplot() + 
     	theme_classic() +
     	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
     	      aspect.ratio = 1) +
     	theme(plot.title = element_text(hjust=0.5, face="bold")) +
     	ggtitle("NCells vs NGenes")
ggsave(filename=paste0(figuresdir, "/", "boxplot_genes_per_cell_final",
                       ".png"), width = 5, 
       height = 5, units = 'in', dpi = 300)
```

The results from the box plot above shows that more transcripts are detected in the PBMC group compared to the rest. This could mean a higher sequencing depth in favour of PBMCs. This will be handled during normalisation.

### **Calculate the number of genes detected per cell in each group.**

```{r nGenes}
seurat@meta.data %>% 
     	ggplot(aes(color=group, x=nFeature_RNA, fill= group)) + 
     	geom_density(alpha = 0.2) + 
     	scale_x_log10() + 
     	theme_classic() +
     	ylab("Cell density") +
     	geom_vline(xintercept = 250)
ggsave(filename=paste0(figuresdir, "/", "UMI_per_cell_final",
                       ".png"), width = 5, 
       height = 5, units = 'in', dpi = 300)
```

We can see that most of the cells have between 800 and 1000 genes detected. It is however worth noticing that that the number of genes detected per cell is higher in the PBMCs compared to LM and CRC conditions.

### **Calculate relation between number of genes, number of transcripts and mitochondrial ratio per cell and group.**

```{r nUMI_gene}
seurat@meta.data %>% 
     	ggplot(aes(x=nFeature_RNA, y=nCount_RNA, color=mitoRatio)) + 
     	geom_point() + 
   	scale_colour_gradient(low = "gray90", high = "black") +
     	stat_smooth(method=lm) +
     	scale_x_log10() + 
     	scale_y_log10() + 
     	theme_classic() +
     	geom_vline(xintercept = 500) +
     	geom_hline(yintercept = 250) +
     	facet_wrap(~group) +
      theme(aspect.ratio = 1)
ggsave(filename=paste0(figuresdir, "/", "mitoratio_genes_and_UMI_final",
                       ".png"), width = 5, 
       height = 5, units = 'in', dpi = 300)
```

When we look at the four quadrants of each group, we see that the upper left quadrant of the cells from the CRC and LM samples, have cells that have a high amount of transcripts but a low number of genes detected. They also have very high percentages of their genes being of mitochondrial origin. This might reflect dead or dying cells. But we also see that the cells have high nUMI counts, so do not meet the criteria of dying cells, that usually have low transcript numbers.  It might also be a cell type, that has a high use of mitochondrial activity and have very few active genes that code for a large number of transcripts. The phenomenon described above is visibly is absent in the PBMC sample. These inconclusive observations will be kept in mind during subsequent analyses.

### **Visualise the mitochondrial counts ratio.**

```{r Mitochondrial counts ratio, echo=FALSE}
# Visualize the distribution of mitochondrial gene expression detected per cell
seurat@meta.data %>% 
     	ggplot(aes(color=group, x=mitoRatio, fill=group)) + 
     	geom_density(alpha = 0.2) + 
     	scale_x_log10() + 
     	theme_classic() +
     	geom_vline(xintercept = 0.1) +
      theme(aspect.ratio = 1)
ggsave(filename=paste0(figuresdir, "/", "mitogenes_per_cell_final",
                       ".png"), width = 5, 
       height = 5, units = 'in', dpi = 300)

``` 

The results show a unimodal distribution of mitochondrial ratios per cell within each group. Intuitively this should reflect reduced number of dead cells. If we had a large number of dead cells, we would have a bimodal distribution of mitochondrial ratio counts. I will select all cells with mitoRatio < 0.1.

### Calculate and visualise complexity.

This metric is to assess the ratio of the number of genes to transcripts. it gives an idea about complexity. Do we have a few very active genes that code for a high number of transcripts? If it is the case, are the less complex cell such as RBCs or may be some cell type that suffered from faulty sequencing or library preparation process.

```{r Complexity}
# Add number of genes per UMI for each cell to metadata
seurat$log10GenesPerUMI <- log10(seurat$nFeature_RNA) / log10(seurat$nCount_RNA)

# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
seurat@meta.data %>%
     	ggplot(aes(x=log10GenesPerUMI, color = group, fill=group)) +
     	geom_density(alpha = 0.2) +
     	theme_classic() +
     	geom_vline(xintercept = 0.8)
ggsave(filename=paste0(figuresdir, "/", "complexity_of_gene_expression_final",
                       ".png"), width = 5, 
       height = 5, units = 'in', dpi = 300)
```

 Most of the cells are highly complex, i.e. complexity score greater than 0.8. All samples practically have their peak in the same location.
 
 
### **Cell filtering to remove low quality cells.**

```{r cell-level filtering}
# inspect metadata
head(seurat@meta.data)

# Filter out low quality reads using selected thresholds - these will change with experiment
filtered_seurat <- subset(x = seurat, 
                         subset= (nCount_RNA >= 500) & 
                           (nFeature_RNA >= 250) & 
                           (log10GenesPerUMI > 0.80) & 
                           (mitoRatio < 0.10))

dim(filtered_seurat) # 22657 123207

saveRDS(filtered_seurat, "rawData/cell_filtered_seurat.rds",
        compress = FALSE)

rm(seurat)
gc()
``` 


### **Gene filtering**

#### **Visualise the most abundant gene families.**

```{r gene_families}
# Compute the relative expression of each gene per cell
png(paste0(figuresdir, "/", "pct_genes_per_biotype", ".png"),
     height = 8, width = 8, units = "in", res = 300)
par(mfrow = c(1, 3),
    mar   = c(4, 6, 1, 1))
rel_expression <-
  Matrix::t(Matrix::t(filtered_seurat@assays$RNA@counts) /
            Matrix::colSums(filtered_seurat@assays$RNA@counts)) * 100
most_expressed <-
  sort(Matrix::rowSums(rel_expression), TRUE) / ncol(filtered_seurat)

# Plot the relative expression of each gene per cell
par(mfrow = c(1, 3),
    mar   = c(4, 6, 1, 1))
boxplot(as.matrix(Matrix::t(rel_expression[names(most_expressed[30:1]), ])),
        cex        = .1,
        las        = 1,
        xlab       = "% total count per cell",
        col        = scales::hue_pal()(90)[30:1],
        horizontal = TRUE,
        ylim       = c(0, 8))
boxplot(as.matrix(Matrix::t(rel_expression[names(most_expressed[60:31]), ])),
        cex        = .1,
        las        = 1,
        xlab       = "% total count per cell",
        col        = scales::hue_pal()(90)[60:31],
        horizontal = TRUE,
        ylim       = c(0, 8))
boxplot(as.matrix(Matrix::t(rel_expression[names(most_expressed[90:61]), ])),
        cex        = .1,
        las        = 1,
        xlab       = "% total count per cell",
        col        = scales::hue_pal()(90)[90:61],
        horizontal = TRUE,
        ylim       = c(0, 8))
dev.off()
```

The results show that some genes constitute up to 10-30% of the counts from a single cell and the other top genes are mitochondrial and ribosomal genes. It is quite common that nuclear lincRNAs have correlation with quality and mitochondrial reads. Let us assemble some information about such genes, which are important for quality control and downstream filtering.
These genes can serve several purposes in single-cell data analysis, such as computing cell quality metrics (see below), normalize data (see below) and even help account for batch effects <https://www.pnas.org/content/116/20/9775>

#### ** % Gene biotype and chromosome location.**

In RNA-sequencing, genes can be categorized into different groups depending on their RNA biotype. For example, “coding”, “non-coding”, “VDJ region genes” are “small interfering RNA” common gene biotypes. Besides, having information about chromosomal location might be useful to identify batch effects driven by sex chromosomes

Depending on the desired type of analysis, some gene categories can be filtered out if not of interest. For single cell specifically, cell libraries are usually constructed using poly-A enrichment and therefore enriching for “protein-coding proteins”, which usually constitutes around 80-90% of all available genes.

```{r gene_biotype}
# Retrieve mouse gene annotation from ENSEMBL
library(biomaRt)

mart = biomaRt::useMart(
  biomart = "ensembl",
  dataset = "hsapiens_gene_ensembl",
  host    = "https://www.ensembl.org")

# Retrieve the selected attributes mouse gene annotation
annot <- biomaRt::getBM(
  mart       = mart,
  attributes = c("external_gene_name", "gene_biotype", "chromosome_name"))

# Match the gene names with theit respective gene biotype
item <- annot[match(rownames(filtered_seurat@assays$RNA@counts), annot[, 1]), "gene_biotype"]
item[is.na(item)] <- "unknown"

# Match the gene names with their respective chromosome location
item <- annot[match(rownames(filtered_seurat@assays$RNA@counts), annot[, 1]),
              "chromosome_name"]
item[is.na(item)] <- "unknown"
item[!item %in% as.character(c(1:23, "X", "Y", "MT"))] <- "other"
``` 

#### **Filter and keep only the protein coding genes. **

```{r protein_coding}
# filter out genes that are non-protein coding

dim(filtered_seurat) # 22657 123207
sel <- annot[match(rownames(filtered_seurat), annot[, 1]), 2] == "protein_coding"
genes_use <- rownames(filtered_seurat)[sel]
genes_use <- as.character(na.omit(genes_use))
filtered_seurat <- filtered_seurat[genes_use, ]
dim(filtered_seurat) # 15820 123207

# save filtered seurat object.
saveRDS(filtered_seurat, 
        "rawData/filtered_seurat.rds",
        compress = FALSE)
rm(rel_expression)
gc()
```

### **Cell cycle scoring and clustering without integration.**

It is recommended to check the cell cycle phase before performing the sctransform method. Since the counts need to be comparable between cells and each cell has a different number of total UMIs, we do a rough normalization by dividing by total counts per cell and taking the natural log. This method isn’t as accurate as the sctransform method that we will use ultimately to identify cell clusters, but it is sufficient to explore sources of variation in our data.

We here perform cell cycle scoring. To score a gene list, the algorithm calculates the difference of mean expression of the given list and the mean expression of reference genes. To build the reference, the function randomly chooses a bunch of genes matching the distribution of the expression of the given list. Cell cycle scoring with Seurat adds three slots in data, a score for S phase, a score for G2M phase and the predicted cell cycle phase. The Seurat package provides a list of human G2M and S phase genes in cc.genes.

```{r cell_cycle}
# perform normal normalisation
filtered_seurat_phase <- NormalizeData(filtered_seurat)

filtered_seurat_phase <- CellCycleScoring(
  object = filtered_seurat_phase,
  g2m.features = cc.genes$g2m.genes,
  s.features = cc.genes$s.genes)

# identify most variable genes
# Identify the most variable genes
filtered_seurat_phase <- FindVariableFeatures(filtered_seurat_phase, 
                                               selection.method = "vst",
                                               nfeatures = 5000, 
                                               verbose = FALSE)

# Scale the counts
filtered_seurat_phase <- ScaleData(filtered_seurat_phase)

# Perform PCA
filtered_seurat_phase <- RunPCA(filtered_seurat_phase, npcs = 20)

# Plot the PCA colored by cell cycle phase
DimPlot(filtered_seurat_phase,
        reduction = "pca",
        group.by= "Phase",
        split.by = "Phase")
ggsave(filename=paste0(figuresdir, "/", "cell_cycle",
                       ".png"), width = 5, 
       height = 5, units = 'in', dpi = 300)
```

Results show that there might be differences due to cell cycle. A cluster is present at G1 phase, which is apparently absent or diminished at G2M and S phases.

```{r unintegrated_clustering}

ElbowPlot(filtered_seurat_phase)

filtered_seurat_phase <- RunUMAP(filtered_seurat_phase,
                                               dims = 1:20,
                                               reduction = "pca")

filtered_seurat_phase <- FindNeighbors(filtered_seurat_phase, dims = 1:2)

filtered_seurat_phase <- FindClusters(filtered_seurat_phase, resolution = 0.5)

# If you haven't installed UMAP, you can do so via 
# reticulate::py_install(packages =
# 'umap-learn')
library(randomcoloR)

set.seed(143)
palette <- distinctColorPalette(100)

# from the plot below, batches are based on the subject id. So I integrate based on subject.ids
DimPlot(filtered_seurat_phase, group.by = "group", order = T)
ggsave(filename=paste0(figuresdir, "/", "groupwise_umap_unintegrated",
                       ".png"), width = 7, 
       height = 7, units = 'in', dpi = 300)
```

Unsupervised clustering and overlaying of the group conditions, shows that there could be some batch effects. This is evidenced by the fact that some cells cluster separately based on their tissue of origin. There are also some clusters with overlaps. CRC, LM and PBMC are different tissues. This might also explain the differences in clustering. In any ways, it will be beneficial to do data integration to see if the differences in clustering is biological or technical.

### **to see if I need to regress out for mitochondrial genes**

```{r mitochondrial gene assessment and visualisation of most varied genes}
# Check quartile values
summary(filtered_seurat_phase@meta.data$mitoRatio)

# Turn mitoRatio into categorical factor vector based on quartile values
filtered_seurat_phase@meta.data$mitoFr <- cut(filtered_seurat_phase@meta.data$mitoRatio, 
                   breaks=c(-Inf, 0.02272, 0.03032 , 0.03329, Inf), 
                   labels=c("Low","Medium","Medium high", "High"))

# Plot the PCA colored by mitoFr
DimPlot(filtered_seurat_phase,
        reduction = "pca",
        group.by= "mitoFr",
        split.by = "mitoFr")
ggsave(filename=paste0(figuresdir, "/", "mito_Fract",
                       ".png"), width = 7, 
       height = 7, units = 'in', dpi = 300)
```

The plot above shows that mitochondrial genes appear evenly distributed across the groups of interest. But at one of the earlier stages, there was a cell population with a high amount of mitochondrial genes. For this reason, i will like to regress out for mitochondrial genes, to avoid a situation where they influence the clustering. 

```{r varies_genes}
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(filtered_seurat_phase), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(filtered_seurat_phase)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
ggsave(filename=paste0(figuresdir, "/", "most_varied_genes",
                       ".png"), width = 7, 
       height = 7, units = 'in', dpi = 300)
```

The figures shows genes related to the fibroblast identity are the most varied genes. Heat shock proteins which are often a sign of cell stress, do not appear at the top 10 most varied genes. It also shows that these cells are in relatively good health and were well processed.

## **Normalisation**

I will sctransform-normalise after splitting the data in to respective libraries, which should contribute to a batch effect. I 
I will regress out the effects of cell cycle and mitochondrial fraction as evidenced by analysis before this level.

Scaling and centering assuming a poisson distribution might in some cases overfit the data, see above. One can overcome this by pooling information across genes with similar abundances in order to obtain more stable parameter estimates to be used as gene weights in the regression model. This is called “scTransform” and, in simple terms, is performing a gene-wise GLM regression using a constrained negative binomial model.

I will use a ‘for loop’ to run the NormalizeData(), CellCycleScoring(), and SCTransform() on each sample, and regress out mitochondrial expression by specifying in the vars.to.regress argument of the SCTransform() function.

Before I run this for loop, I know that the output can generate large R objects/variables in terms of memory. If I have a large dataset, then I might need to adjust the limit for allowable object sizes within R (Default is 500 * 1024 ^ 2 = 500 Mb) using the following code:

```{r sct_normalisation}
# increase amount of allowable subjects in R
options(future.globals.maxSize = 4000 * 1024^2)#

split_seurat <- SplitObject(filtered_seurat, split.by = "group")

unique(filtered_seurat$group)

split_seurat <- split_seurat[unique(filtered_seurat$group)]

for (i in 1:length(split_seurat)) {
  
    split_seurat[[i]] <- NormalizeData(split_seurat[[i]], verbose = TRUE)
    split_seurat[[i]] <- CellCycleScoring(split_seurat[[i]], 
                                          g2m.features= cc.genes.updated.2019$g2m.genes, 
                                          s.features=cc.genes.updated.2019$s.genes)
    split_seurat[[i]]$CC.Difference <- split_seurat[[i]]$S.Score - split_seurat[[i]]$G2M.Score
    split_seurat[[i]] <- SCTransform(split_seurat[[i]], variable.features.n = 5000,
                                     vars.to.regress = c("CC.Difference", "mitoRatio"))
}
```

## Integration

I perform integration across the **group** variable where I logically expect batch effects. This integration method makes use of mutual nearest neighbour. I want to identify comparable cell types and if possible present in all subjects.

Integration is a powerful method that uses these shared sources of greatest variation to identify shared subpopulations across conditions groups <https://www.sciencedirect.com/science/article/pii/S0092867419305598?via%3Dihub> . The goal of integration is to ensure that the cell types of one condition/dataset align with the same celltypes of the other conditions/datasets (e.g. control macrophages align with IPF macrophages). But in this data set, I expect same cells to cluster differently depending on their cncerous state.

```{r integration}
# Select the most variable features to use for integration
integ_features <- SelectIntegrationFeatures(object.list = split_seurat, 
                                            nfeatures = 5000) 

# Prepare the SCT list object for integration
split_seurat <- PrepSCTIntegration(object.list = split_seurat, 
                                   anchor.features = integ_features)

# do pca before finding integration achors
split_seurat <- lapply(X = split_seurat, FUN = RunPCA, 
                       verbose = FALSE, features = integ_features)

# Find best buddies - can take a while to run
integ_anchors <- FindIntegrationAnchors(object.list = split_seurat, 
                                        normalization.method = "SCT", dims = 1:50,
                                        anchor.features = integ_features,
                                        reduction = "rpca")
# integrate across libraries
seurat_integrated <- IntegrateData(anchorset = integ_anchors, 
                                   dims = 1:50,
                                   normalization.method = "SCT", verbose = T)
saveRDS(seurat_integrated,
        paste0(objectsdir, 
                         "/","seurat_filtered_and_integrated", 
                         ".rds"),
        compress = FALSE)

rm(integ_anchors, seurat_rna, split_seurat, seurat)
gc()
```

After performing the integration, the next step is to do dimensionality reduction.

## **Dimensionality reduction.**

PCA helps find the most varied features in the data set.

```{r PCA}
# Run PCA
seurat_integrated <- RunPCA(object = seurat_integrated, npcs=100)

# Plot PCA
PCAPlot(seurat_integrated,
        dims = c(1, 2),
        split.by = "group") 
ggsave(filename=paste0(figuresdir, "/", "PCA_overlay_of_groups_integrated",
                       ".png"), width = 5, 
       height = 5, units = 'in', dpi = 300)

# Explore heatmap of PCs
DimHeatmap(seurat_integrated, 
           dims = 1:9, 
           cells = 500, 
           balanced = TRUE)
ggsave(filename=paste0(figuresdir, "/", "heatmap_PCs_integrated",
                       ".png"), width = 5, 
       height = 5, units = 'in', dpi = 300)

# Printing out the most variable genes driving PCs
print(x = seurat_integrated[["pca"]], 
      dims = 1:10, 
      nfeatures = 5)

# Plot the elbow plot
ElbowPlot(object = seurat_integrated, 
          ndims = 100)

# get best number of PCs
# we can calculate the elbow by
# 1- The point where the principal components only contribute 5% of standard deviation and the principal components cumulatively contribute 90% of the # standard deviation.
# 2 - The point where the percent change in variation between the consecutive PCs is less than 0.1%.

# Determine percent of variation associated with each PC
pct <- seurat_integrated[["pca"]]@stdev / sum(seurat_integrated[["pca"]]@stdev) * 100

# Calculate cumulative percents for each PC
cumu <- cumsum(pct)

# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
co1 <- which(cumu > 90 & pct < 5)[1]
co1

# Determine the difference between variation of PC and subsequent PC
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1

# last point where change of % of variation is more than 0.1%.
co2

# Minimum of the two calculation
pcs <- min(co1, co2)
pcs # 14
```

## Clustering

```{r clustering}
# Run UMAP
seurat_integrated <- RunUMAP(seurat_integrated, 
                             dims = 1:co1,
			                       reduction = "pca")

# Determine the K-nearest neighbor graph
seurat_integrated <- FindNeighbors(object = seurat_integrated, 
                                dims = 1:2, reduction="umap")
                                
# Determine the clusters for various resolutions                                
seurat_integrated <- FindClusters(object = seurat_integrated,
                               resolution = seq(0.05, 2.4, 0.05),
                               dims = 1:co1)

DimPlot(seurat_integrated)

DefaultAssay(seurat_integrated) <- "RNA"

DimPlot(seurat_integrated, group.by = "group") + theme(aspect.ratio = 1)

# for visualisation purposes
seurat_integrated <- NormalizeData(seurat_integrated)

saveRDS(seurat_integrated,
        paste0(objectsdir, 
                         "/","integrated_and_clustered", 
                         ".rds"))
```

